qir/qat/Passes/GroupingPass/Tests/Unit/grouping.cpp

```diff
@@ -0,0 +1,525 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "gtest/gtest.h"
+#include "qir/qat/AdaptorFactory/ConfigurableQirAdaptorFactory.hpp"
+#include "qir/qat/Llvm/Llvm.hpp"
+#include "qir/qat/Passes/GroupingPass/GroupingPass.hpp"
+#include "qir/qat/Passes/StaticResourceComponent/AllocationAnalysisPass.hpp"
+#include "qir/qat/Passes/StaticResourceComponent/StaticResourceComponentConfiguration.hpp"
+#include "qir/qat/Passes/TargetQisMappingPass/Factory.hpp"
+#include "qir/qat/TestTools/IrManipulationTestHelper.hpp"
+#include "qir/qat/Utils/FunctionToModule.hpp"
+
+#include <functional>
+
+using namespace microsoft::quantum;
+
+namespace
+{
+using IrManipulationTestHelperPtr = std::shared_ptr<IrManipulationTestHelper>;
+IrManipulationTestHelperPtr newIrManip(std::string const& script)
+{
+    IrManipulationTestHelperPtr ir_manip = std::make_shared<IrManipulationTestHelper>();
+
+    ir_manip->declareOpaque("Qubit");
+    ir_manip->declareOpaque("Result");
+    ir_manip->declareOpaque("Array");
+    ir_manip->declareOpaque("Tuple");
+    ir_manip->declareOpaque("Range");
+    ir_manip->declareOpaque("Callable");
+    ir_manip->declareOpaque("String");
+
+    ir_manip->declareFunction("i64 @get_arg()");
+    ir_manip->declareFunction("void @no_opt(i64)");
+
+    ir_manip->declareFunction("void @__quantum__rt__array_start_record_output()");
+    ir_manip->declareFunction("void @__quantum__rt__array_end_record_output()");
+    ir_manip->declareFunction("void @__quantum__rt__int_record_output(i64)");
+
+    ir_manip->declareFunction("%Qubit* @__quantum__rt__qubit_allocate()");
+    ir_manip->declareFunction("void @__quantum__rt__qubit_release(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__qis__h__body(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__qis__z__body(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__qis__y__body(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__qis__x__body(%Qubit*)");
+    ir_manip->declareFunction("i8* @__quantum__rt__array_get_element_ptr_1d(%Array*, i64)");
+
+    ir_manip->declareFunction("void @__quantum__qis__cnot__body(%Qubit*, %Qubit*)");
+    ir_manip->declareFunction("%Result* @__quantum__qis__m__body(%Qubit*)");
+    ir_manip->declareFunction("%Result* @__quantum__rt__result_get_zero()");
+    ir_manip->declareFunction("void @__quantum__rt__result_record_output(%Result*)");
+    ir_manip->declareFunction("void @__quantum__rt__result_update_reference_count(%Result*, i32)");
+    ir_manip->declareFunction("void @__quantum__rt__qubit_release(%Qubit*)");
+    ir_manip->declareFunction("void @__quantum__rt__bool_record_output(i1)");
+    ir_manip->declareFunction("void @__quantum__rt__array_start_record_output()");
+    ir_manip->declareFunction("void @__quantum__rt__array_end_record_output()");
+
+    ir_manip->declareFunction("void @__quantum__qis__mz__body(%Qubit*, %Result*)");
```

Comment:
I am taking it the grouping does not rely on attributes at this time and instead goes by name or signature? We are going to attach an attribute to qis functions that are irreversible (see the Base Profile spec). It would be nice for the grouping to go by that attribute, but I am good as is - that shouldn't be too hard to tweak.

Reply:
As it stands, the code uses a predefined list in the config. The intension is to extend the list using attributes, but the implementation is not there yet.